### ISSUE TITLE
Tool to generate "Configuration Reference" Table for applications

### ISSUE THREAD

**Issue by wansors** [quarkusio/quarkus#19020]
### Description

In large projects there are lots of configuration properties, and is not easy to keep up to date the documentation.
It could be nice to have a tool or maven plugin to autogenerate a "Configuration Reference" Table like the ones on the website (https://quarkus.io/guides/all-config). Adding the property, documentation from the javadoc and default values.

Thanks

**Comment by famod** [comment]
Do you mean some kind of filtered config based on the extensions you are using?

**Comment by wansors** [comment]
> Do you mean some kind of filtered config based on the extensions you are using?

Not of the extensions activated, only for the new config properties added by the user in their project. It has no sense for small applications, but if your application or service has many configuration parameters is a nice way to have docslikecode.

**Comment by faermanj** [comment]
This would be extremely useful in our projects.

**Comment by radcortez** [comment]
This is possible by adding the `quarkus-extension-processor` as a dependency tp the project:

```xml
<dependency>
  <groupId>io.quarkus</groupId>
  <artifactId>quarkus-extension-processor</artifactId>
  <version>QUARKUS_VERSION</version>
</dependency>
```

Note that the processor dependency is not managed by the BOM, so it requires to set the version.

The `@ConfigMapping` interfaces, also require an additional annotation for the processor to discover them, used to define the config phase (unique to Quarkus), but no real impact in the application code:

```java
@ConfigRoot(phase = ConfigPhase.RUN_TIME)
@ConfigMapping(prefix = "my.config")
public interface MyConfig {
}
```

In this way, the build will start to generate `ascidoc` in the build tool output folder. Also, it will fail the build if a mapping method is missing the javadoc.

**Comment by michalsomora** [comment]
Hi
would it be possible to generate documentation for config with internal classes as a flat structure?
So for instance this config mapping generate ascidoc only on 1st level, e.g. `my.config.internal` but not for `my.config.internal.field1`
To do so, i need to extract InternalConfig to separate class with its own prefix.
`
<pre>
@ConfigRoot(phase = ConfigPhase.RUN_TIME)
@ConfigMapping(prefix = "my.config")
public interface MyConfig {
    /**
     * internal doc
     * @return
     */
    @WithName("internal")
    InternalConfig internal();
    interface InternalConfig {
        /**
         * field1 doc
         * @return
         */
        @WithName("field1")
        String field1();
    }
}
</pre>`
 It would be convenient to have all the configs under one class
Thanks for any advice.

**Comment by radcortez** [comment]
Which Quarkus version are you using? I believe we had a bug in that case, but it was fixed some time ago. I've tried your example, and I've got the expected result:

![Screenshot 2024-01-23 at 13 28 23](https://github.com/quarkusio/quarkus/assets/5796305/b49bdb8d-283d-4773-835c-6ca855307b0a)

**Comment by michalsomora** [comment]
Hi @radcortez 
thanks for the quick reply, i am using 2.16.2.Final, probablly its not fixed there

**Comment by radcortez** [comment]
Yes, I think it is only fixed in 3. something. You may want to try to annotate the nested group with `@ConfigGroup`. It may help for that version (but I'm not sure).

**Comment by gsmet** [comment]
I added that it would be nice to have a plugin specific for doc generation separate from the extension plugin for applications in the issue here: https://github.com/quarkusio/quarkus/issues/35439 .

**Comment by Athou** [comment]
> I added that it would be nice to have a plugin specific for doc generation separate from the extension plugin for applications in the issue here: #35439 .

Hi @gsmet,

It looks like https://github.com/quarkusio/quarkus/issues/35439 has been completed in Quarkus 3.14. Does that mean there is now a way to generate application configuration documentation? 

Thanks!

**Comment by gsmet** [comment]
This is actually an issue we should probably reopen. The extension annotation processor was never designed to handle applications but it apparently worked by accident and I broke it by making assumptions about it being run in an extension (which doesn't look so far fetched given it's the extension annotation processor :)).

I really think we should have a different annotation processor though, that could share some of the code (we need to make really sure it won't make the extension annotation processor harder to maintain though).

**Comment by gsmet** [comment]
@Athou were you actually using it before or are you trying to use it now?

**Comment by gsmet** [comment]
Ah cool, then I have a project to experiment with. I'll see if I can unblock you.

Where do you consume this output? Because I don't think we use anything special for the icons, you might need font awesome around though.

**Comment by gsmet** [comment]
@Athou I thought given your message in the PR that it wasn't working for you anymore but it's actually working, right?

**Comment by gsmet** [comment]
Also I'm wondering if it would help if there was an option to generate Markdown?

**Comment by Athou** [comment]
> @Athou I thought given your message in the PR that it wasn't working for you anymore but it's actually working, right?

Yes, it works with Quarkus 3.14. In addition to the `quarkus-extension-processor`  I had to add 

```xml
<plugin>
	<groupId>io.quarkus</groupId>
	<artifactId>quarkus-config-doc-maven-plugin</artifactId>
	<version>${quarkus.version}</version>
	<extensions>true</extensions>
	<executions>
		<execution>
			<id>default-generate-asciidoc</id>
			<phase>process-test-resources</phase>
			<goals>
				<goal>generate-asciidoc</goal>
			</goals>
		</execution>
	</executions>
</plugin>
```
because the processor no longer generates asciidoc. I had to add an execution on another phase because the default phase executes before the processor is invoked.

The only issue I currently have is the one I mentioned in https://github.com/quarkusio/quarkus/pull/42561#issuecomment-2316749444, I'm no longer able to run my unit tests in IntelliJ but the maven build works fine.

> Where do you consume this output? Because I don't think we use anything special for the icons, you might need font awesome around though.

I currently don't. I just version the output of  `quarkus-config-doc-maven-plugin` and I wrote a [unit test](https://github.com/Athou/commafeed/blob/994f1fb121cff2cf5ea03314cafe5485c7f89c16/commafeed-server/src/test/java/com/commafeed/CommaFeedConfigurationTest.java) that verifies that the asciidoc is up to date. Then I just use the asciidoc renderer of Github because it's output is good enough. I should probably transform the asciidoc source to html and publish the result to Github pages though.

> Also I'm wondering if it would help if there was an option to generate Markdown?

That would be awesome, yes :)

**Comment by gsmet** [comment]
>> Also I'm wondering if it would help if there was an option to generate Markdown?

> That would be awesome, yes :)

Would you be interested in collaborating on that?

Basically, we would just need to have a Markdown version of these templates:
https://github.com/quarkusio/quarkus/tree/main/devtools/config-doc-maven-plugin/src/main/resources/templates

And a Markdown version of this class:
https://github.com/quarkusio/quarkus/blob/main/devtools/config-doc-maven-plugin/src/main/java/io/quarkus/maven/config/doc/AsciidocFormatter.java
(probably with a common parent class)

And then refactor the Mojo to have a parent class and have a second mojo for Markdown.

Even if you just help with translating the templates to Markdown, that would help a lot.

Maybe I could prepare the infrastructure and you could play with the templates to make them work properly?

**Comment by gsmet** [comment]
As for your issue with Intellij, https://github.com/quarkusio/quarkus/pull/42853 should take care of it!

**Comment by radcortez** [comment]
Also, as a workaround, you may be able to point the processor to the working version (while leaving Quarkus with the new version).

**Comment by Athou** [comment]
> > > Also I'm wondering if it would help if there was an option to generate Markdown?
> 
> > That would be awesome, yes :)
> 
> Would you be interested in collaborating on that?
> 
> Basically, we would just need to have a Markdown version of these templates: https://github.com/quarkusio/quarkus/tree/main/devtools/config-doc-maven-plugin/src/main/resources/templates
> 
> And a Markdown version of this class: https://github.com/quarkusio/quarkus/blob/main/devtools/config-doc-maven-plugin/src/main/java/io/quarkus/maven/config/doc/AsciidocFormatter.java (probably with a common parent class)
> 
> And then refactor the Mojo to have a parent class and have a second mojo for Markdown.
> 
> Even if you just help with translating the templates to Markdown, that would help a lot.

Sure, I can take a crack at it.

> Maybe I could prepare the infrastructure and you could play with the templates to make them work properly?

That'd be very helpful, thanks!

> As for your issue with Intellij, https://github.com/quarkusio/quarkus/pull/42853 should take care of it!

Thank you!

**Comment by gsmet** [comment]
OK, I'll prepare the necessary infrastructure. In the meantime, you can play with the Qute templates as it will be needed in any case and we can assemble all of this once everything is ready.

**Comment by gsmet** [comment]
@Athou see https://github.com/quarkusio/quarkus/pull/42907 for a good start on this. Ping me in the PR if you need help iterating to get the Markdown templates working.
